### ISSUE TITLE
Extremely slow update_actor in GRPO with Qwen3-4B + FSDP2 on Ascend 910B

### ISSUE THREAD

**Issue by SsllyZz** [verl-project/verl#6010]
### System Info

Hi, thanks for the great work on VERL.

I’m training **Qwen3-4B** using **GRPO + FSDP2** on **8× Ascend 910B**, and observed very slow training:

* **~800 seconds per step**
* `update_actor` dominates the step time

### Key Config

* `max_prompt_length=4096`
* `max_response_length=8192`
* `rollout.n=8`
* `actor.ppo_micro_batch_size_per_gpu=4`
* `actor.use_dynamic_bsz=True`
* FSDP2 with `fsdp_size=8`

### Observation

`update_actor` is significantly slower than other stages and becomes the main bottleneck. The overall throughput is much lower than expected for a 4B model on 8 GPUs.

### Questions

1. Is this expected for **Qwen3 + GRPO + FSDP2** on Ascend?
2. Are there known performance bottlenecks in `update_actor` under this setup?
3. Any recommended configs (e.g., token budget, batch sizing) to improve actor update throughput?

I can provide full logs and profiling if needed.

Thanks!

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

Below is a simplified reproduction config.

### Environment

* 8 × Ascend 910B
* VERL + GRPO
* Qwen3-4B
* `vllm==0.18.0`
* `vllm-ascend@54879467`
* `transformers@cc7ab9be`

### Launch script

```bash
export RAY_TMPDIR=/mnt1/sfs_turbo/maoyingzhi2024/tmp
export GLOO_SOCKET_IFNAME=enp23s0f3
export NCCL_SOCKET_IFNAME=enp23s0f3

python3 -m verl.trainer.main_ppo \
    algorithm.adv_estimator=grpo \
    data.train_batch_size=128 \
    data.max_prompt_length=4096 \
    data.max_response_length=8192 \
    actor_rollout_ref.model.path=/mnt1/Qwen3.5-4B \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.actor.ppo_mini_batch_size=32 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=4 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.01 \
    actor_rollout_ref.actor.use_torch_compile=False \
    actor_rollout_ref.actor.strategy=fsdp2 \
    actor_rollout_ref.ref.strategy=fsdp2 \
    actor_rollout_ref.actor.fsdp_config.fsdp_size=8 \
    actor_rollout_ref.actor.fsdp_config.reshard_after_forward=True \
    actor_rollout_ref.ref.fsdp_config.reshard_after_forward=True \
    actor_rollout_ref.actor.fsdp_config.entropy_checkpointing=True \
    actor_rollout_ref.actor.entropy_from_logits_with_chunking=True \
    actor_rollout_ref.actor.use_dynamic_bsz=True \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=1 \
    actor_rollout_ref.ref.fsdp_config.param_offload=False \
    actor_rollout_ref.ref.use_torch_compile=False \
    actor_rollout_ref.rollout.name=vllm \
    actor_rollout_ref.rollout.tensor_model_parallel_size=2 \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.6 \
    actor_rollout_ref.rollout.n=8 \
    actor_rollout_ref.rollout.enable_chunked_prefill=True \
    actor_rollout_ref.rollout.max_num_batched_tokens=4096 \
    actor_rollout_ref.rollout.free_cache_engine=True \
    actor_rollout_ref.rollout.enforce_eager=True \
    actor_rollout_ref.rollout.enable_prefix_caching=False \
    actor_rollout_ref.rollout.max_model_len=8192 \
    trainer.n_gpus_per_node=8 \
    trainer.nnodes=1 \
    trainer.total_epochs=2
```

### Expected behavior

For a similar setup (same hardware and comparable configs), Qwen3 models achieve ~100s per step, and update_actor does not dominate the training time.

Therefore, I would expect Qwen3-4B under GRPO + FSDP2 to have comparable performance, rather than ~800s/step with update_actor as the bottleneck.

**Comment by WangYijun-OUC** [comment]
same problem

**Comment by burnningwings** [comment]
> same problem

+1

**Comment by Kirrito-k423** [comment]
Thank you for the detailed performance report. The ~800 seconds per step with `update_actor` dominating is indeed concerning for a 4B model on 8 GPUs.

**Potential causes for slow `update_actor` on Ascend NPU**:

1. **FSDP2 + Ascend compatibility**: FSDP2 was primarily developed for NVIDIA GPUs. Ascend NPU may have different memory allocation patterns that cause inefficiencies in FSDP2's sharding strategy.

2. **Dynamic batch size with long sequences**: Your `max_response_length=8192` combined with `use_dynamic_bsz=True` and `ppo_micro_batch_size_per_gpu=4` may create large gradient computation overhead. Try:
   - Reduce `max_response_length` to 4096
   - Increase `ppo_micro_batch_size_per_gpu` to 8 or 16 (if memory allows)

3. **Gradient checkpointing overhead**: `entropy_checkpointing=True` + `entropy_from_logits_with_chunking=True` add computation overhead. Consider disabling if memory is sufficient.

4. **Communication bottleneck**: With `fsdp_size=8`, all 8 GPUs participate in FSDP2 synchronization. On Ascend, the HCCL communication may have different latency characteristics compared to NCCL.

**Recommended diagnostics**:

1. **Profile `update_actor`**: Use Ascend profiling tools to identify hotspots:
   ```bash
   # Enable profiling
   export PROFILING_MODE=enable
   ```

2. **Compare with FSDP1**: Try `strategy=fsdp` instead of `strategy=fsdp2` to see if performance differs.

3. **Reduce token budget**: Your `rollout.n=8` with `max_response_length=8192` generates 8 × 8192 = 65K tokens per sample. Consider:
   - `rollout.n=4`
   - `max_response_length=4096`

4. **Related PRs**: Check PR [#6089](https://github.com/verl-project/verl/pull/6089) which fixes `routed_experts` truncation for MoE models - this could affect Qwen3 performance.

**Alternative backend**: Consider using VeOmni backend for Ascend NPU - see PR [#6061](https://github.com/verl-project/verl/pull/6061) which adds Ascend-specific optimizations.

Could you share:
- The exact Ascend profiling output if available?
- Whether `fsdp` (FSDP1) performs differently?
- The `update_actor` time breakdown (forward/backward/optimizer)?

**Comment by wucong25** [comment]
Qwen3-4B or Qwen3.5-4B?

**Comment by leetakhing** [comment]
Hi @SsllyZz ,

Thanks a lot for sharing this detailed issue! I'm running the exact same setup (8× Ascend 910B + Qwen3-4B + GRPO + FSDP2) as you.

I've been stuck on `hccl api failed` errors with all newer VERL commits. The only one that works for me is the 4-month-old `f31df344`, but it's too outdated and has other issues.

Since your environment works, could you share the exact versions you used? I just need:
- VERL commit hash
- vllm-ascend & transformers commits
- torch/torch-npu & CANN toolkit versions
- Any other key dependencies that matter

Would really appreciate your help!
